photon-lib/src/main/scala/com/linkedin/photon/ml/function/PriorDistribution.scala

```diff
@@ -0,0 +1,256 @@
+   * @param coefficients The model coefficients for which to compute the objective function's gradient
+   * @param normalizationContext The normalization context
+   * @return The gradient of the objective function and regularization terms
+   */
+  abstract override protected[ml] def gradient(
+      input: Data,
+      coefficients: Coefficients,
+      normalizationContext: BroadcastWrapper[NormalizationContext]): Vector[Double] =
+    calculate(input, coefficients, normalizationContext)._2
+
+  /**
+   * Compute both the value and the gradient of the function over the given data for the given model coefficients, with
+   * regularization towards the prior coefficients (computing value and gradient at once is more efficient than
+   * computing them sequentially).
+   *
+   * @param input The data over which to compute the objective function value and gradient
+   * @param coefficients The model coefficients for which to compute the objective function's value and gradient
+   * @param normalizationContext The normalization context
+   * @return The value and gradient of the objective function and regularization terms
+   */
+  abstract override protected[ml] def calculate(
+      input: Data,
+      coefficients: Coefficients,
+      normalizationContext: BroadcastWrapper[NormalizationContext]): (Double, Vector[Double]) = {
+
+    val (baseValue, baseGradient) = super.calculate(input, coefficients, normalizationContext)
+    val valueWithRegularization = baseValue + l1RegValue(convertToVector(coefficients)) +
+      l2RegValue(convertToVector(coefficients))
+    val gradientWithRegularization = baseGradient + l1RegGradient(convertToVector(coefficients)) +
+      l2RegGradient(convertToVector(coefficients))
+
+    (valueWithRegularization, gradientWithRegularization)
+  }
+
+  /**
+   * Compute the gradient of the Laplace term for the given model coefficients. Gradient is
+   * l1RegWeight :/ sqrt(priorVariance) if coefficients >= priorMeans;
+   * - l1RegWeight :/ sqrt(priorVariance) if coefficients < priorMeans.
+   *
+   * @param coefficients The model coefficients
+   * @return The gradient of the Laplace regularization term
+   */
+  protected def l1RegGradient(coefficients: Vector[Double]): Vector[Double] = {
+
+    val coefficientsMask = (coefficients - priorMeans).map(coefficient => if (coefficient > 0) 1.0 else -1.0)
+
+    l1RegWeight * (coefficientsMask *:* sqrt(inversePriorVariances))
+  }
+
+  /**
+   * Compute the gradient of the Gaussian regularization term for the given model coefficients. Gradient is
+   * l2RegWeight * (coefficients - priorMeans) :/ priorVariance.
+   *
+   * @param coefficients The model coefficients
+   * @return The gradient of the Gaussian regularization term
+   */
+  protected def l2RegGradient(coefficients: Vector[Double]): Vector[Double] = {
+
+    val normalizedCoefficients = (coefficients - priorMeans) *:* inversePriorVariances
```

Comment:
This is one point of difference that we can discuss.

Reply:
We use the prior model information by adding a new regularization term: (coefficients - priorMeans) ^ 2 / priorVariances * l2RegWeight / 2